bftengine/src/preprocessor/messages/PreProcessResultMsg.cpp

```diff
@@ -57,8 +57,8 @@ std::optional<std::string> PreProcessResultMsg::validatePreProcessResultSignatur
   const auto [buf, buf_len] = getResultSignaturesBuf();
   auto sigs = preprocessor::PreProcessResultSignature::deserializeResultSignatureList(buf, buf_len);
 
-  // f because the primary also executes the request (check PreProcessor::numOfRequiredReplies())
-  const auto expectedSignatureCount = fVal;
+  // f+1 signatures are required
+  const auto expectedSignatureCount = fVal + 1;
   if (sigs.size() < static_cast<uint16_t>(expectedSignatureCount)) {
```

Comment:
we should change sigs.size() < static_cast<uint16_t>(expectedSignatureCount) to sigs.size() != static_cast<uint16_t>(expectedSignatureCount) if we are sure that f+1 signatures will come.
____________________________________________________________________________________
But in case the number of signatures are greater than f+1, then we should add more logic for the following scenario [Read below if there is any chance that sigs.size() > f + 1, else ignore]:
lets say in one PreProcessResultMsg there are 7 signatures from s1, s2, to s7. Here the number of signatures is greater than 3. Here f is 2, so we need 3 correct signatures out of 7. But in case we get 7, then in the verification loop we have to then check is out of 7 atleast 3 have arrived or not. The moment we get 3 correct signatures we should break the loop and return success.
case1: s1, s2 are bad. s3, s4,s5,s6 and s7 are good. The loop should run for s1 to s5. Ignore that s1 and s2 are failed and return success.
case 2: s6 and s7 are bad and rest are good. The loop should run for s1 to s3 and return success.
case 3: s2 and s6 are bad and rest are good. The loop should run for s1 to s4. Returns success.
A failure case will be, when loop don't even get 3 good signatures.

Reply:
This is an interesting point, @arc-vmware. I think you are right - we should check that the message contains exactly f+1 signatures.

In practice the primary considers the PreProcessing completed once it receives f replies with result hash matching its own. So we should always receive f+1 signatures (f from replicas and 1 from primary).

I think the primary should not send invalid signatures at all - this will be an indication for a serious error.

Also a malicious primary can launch a DoS attack against the replicas by sending a PreProcessReply with a huge number of signatures. We shouldn't verify more signatures than we need.

@yuliasherman what do you think? Should check if sigs.size() == expectedSigSize?

Reply:
Agree